rfcs/021 - Datastore Restructuring RFC.md

```diff
@@ -0,0 +1,82 @@
+
+1. Determining the primary node for mutations or consistent reads, and the closest node for fast eventually consistent reads.
+2. Determining all nodes the primary node should replicate to for backup and eventually consistent usage.
+3. Redirecting requests incorrectly sent to the node if the other node erroneously identified it (can happen during scale up/down and network partitioning, silent redirection also handles partial partitioning of the cluster.
+
+If we put the routing layer in front of all updates to the storage layer as well, then the execution layer does not need to worry about this, anymore; and can simply do the work asked of it. The execution layer would look more like the local version implementation in `opcode.rs` that way.
+
+For the storage layer, a pattern we used prior at Uber for high volume mutations should work well here: We create a threadpool (actual OS threadpool) and each thread has its own hashmap internally. We decide which keys go to which hashmap very similarly to how we choose which node is primary for a key, with a RendezvousHash-inspired algorithm, so guarantee it always goes to the same location. Since we are basically guaranteed that the number of CPU cores on any given node is static, however, we can make it a simple `xxhash` of the key `modulus` the number of OS threads to identify the thread we want, which is much faster than a true RendezvousHash (designed to handle changes in cluster membership). Access to the data in these OS threads can be done with channels. We can have three channels, two reads and a write, where one read channel provides a token value that must be passed back through the write channel to let the next locking read have access to the data, etc. Since we allow explicit setting of data in datastore, as well, the write channel will accept writes without a token, but obviously any key with both tokenless and token writes will have data race issues. It would be best if we can adjust the API exposed in the standard library to separate those usages from each other.
+
+It may also make sense for the storage layer to get the execution context, as well; then no transferrance of the key in question between OS threads would be necessary and since blocking mutations like that prevent other mutations on the same key from happening simultaneously, that might be a better use of CPU time on these OS threads, as well. But mixing Tokio async/await with real OS threads may be difficult/impossible, or may still require spawning a task with an `async move` block inside, which would nullify the benefits while complicating the storage layer, so we should be careful with this. Likely we don't do something like this in the first pass to make the layers cleaner and more easily extensible, and we come back to it later as an optimization pass when we're more convinced that little will change with datastore's API and the increased maintenance burden is worth it.
```

Comment:
> or may still require spawning a task with an `async move` block inside, which would nullify the benefits while complicating the storage layer

I'm not sure how the benefits would be nullified? In fact, I think it's better for performance overall without impacting the API at all - the fewer OS threads there are, the less likely any OS thread running the AVM will be unscheduled. Being able to schedule the tokio task on any currently-running OS thread while some other AVM task is running an IO-bound opcode also prevents other OS threads from hogging CPU resources, preventing the hashmap thread from getting run until some other thread yields CPU time.

If this were implemented using OS threads, it would look like:
```rs
thread::spawn(move {
  let mut data = HashMap::new();
  loop {
    if let Ok(req) = read_requests.try_recv() {
      ...
    }
    if let Ok(req) = write_requests.try_recv() {
      ...
    }
    thread::yield_now();
  }
}
```
whereas with tokio, you could do:
```rs
tokio::spawn(async move {
  let mut data = HashMap::new();
  loop {
    // note that `tokio::select` works similarly to rust's `match`, and the
    // requests can be handled directly inside of the `select` instead of
    // returning `(Option<ReadRequest>, Option<WriteRequest>)`
    let (read_req, write_req) = tokio::select! {
      req = read_requests.recv() => (Some(req.unwrap()), None),
      req = write_requests.recv() => (None, Some(req.unwrap())),
    };
    if let Some(read_req) = read_req {
      ...
    } else if let Some(write_req) = write_req {
      ...
    }
  }
});
```
Excluding the `thread::yield_now()` in the OS thread version would result in a hotloop that would infinitely consume CPU resources - this automatically gets handled when `.poll()`ing the `recv()` futures in the tokio version, and doesn't happen until tokio tries several times to receive from each of the channels.

Reply:
So most CPUs are [NUMA](https://en.wikipedia.org/wiki/Non-uniform_memory_access) at least at the CPU cache layer. If the data that was in that hashmap was in cache for the CPU core the OS thread was tied to, then the `async move` for a tokio task is gonna require whichever other thread on whichever other CPU core to instead reacquire the data from RAM where the addressing is uniform again.

If we can make sure that the particular compute task using that data actually executes in a blocking manner on the OS thread that houses the data, we *aren't* "hogging CPU resources", they're being used to actually do the relevant work and immediately stored the result back where it belongs, and then immediately polling the channels for work and executing it, or sleeping for 100ns or something and trying again would be *less* wasted time than using the Tokio framework here.

Reply:
The `async move` only happens once per hashmap though. If you're worried about task affinity, <https://docs.rs/tokio/1.10.0/tokio/task/fn.spawn_local.html> can be used in conjunction with <https://docs.rs/tokio/1.10.0/tokio/runtime/struct.Builder.html#method.on_thread_start> to start a hashmap task *per OS thread*.

Reply:
I think we're talking over each other? I'm not referring to the `async move` on the initialization of the thread (though the task affinity piece is nice to know and would let us use tokio tasks here, too); I was referring to the idea that the datastore remote execution opcodes could run "in place" or if the data is copied into a tokio task and executed elsewhere and then brought back. if my `dsmrun` opcode call just happened in the same thread as where the cached data is located, and the inner opcode calls were all `cpu!` opcodes, then we skip a *ton* of scheduling overhead and memcpys.